subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java

```diff
@@ -233,12 +253,52 @@ private FileLockCommunicator getCommunicator() {
         }
     }
 
-    private static class ContendedAction {
-        private final Runnable action;
+    private class ContendedAction implements Runnable {
+        private final Lock lock = new ReentrantLock();
+        private final long lockId;
+        private final Action<FileLockReleasedSignal> action;
+        private Set<SocketAddress> requesters = new LinkedHashSet<SocketAddress>();
         private boolean running;
 
-        private ContendedAction(Runnable action) {
+        private ContendedAction(long lockId, Action<FileLockReleasedSignal> action) {
+            this.lockId = lockId;
             this.action = action;
         }
+
+        @Override
+        public void run() {
+            action.execute(new FileLockReleasedSignal() {
+                @Override
+                public void trigger() {
+                    Set<SocketAddress> requesters = consumeRequesters();
+                    if (requesters == null) {
+                        throw new IllegalStateException("trigger() has already been called and must at most be called once");
+                    }
+                    communicator.confirmLockRelease(requesters, lockId);
```

Comment:
The confirmation message is identical to the one sent for confirming unlock requests. Thus, old versions interpret it as such and handle it in https://github.com/gradle/gradle/blob/v4.9.0/subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java#L135-L138. Since the processes communicate using all local network addresses, duplicate unlock request confirmations already occur today (e.g. for local IPv4 and IPv6 addresses). If you have an idea how to put this into an automated test, I'd be happy to add it.

Reply:
If it is identical wouldn't that mean that a duplicate request confirmation would be wrongly interpreted as a release confirmation?

Reply:
We could change the data structure to keep track of addresses in addition to ports to improve that, though.

Reply:
Or change the message so we actually know what the other one is trying to tell us. Nothing that blocks this PR, but it seems this is creating unnecessary chatter on the local network.